### ISSUE TITLE
[no-unused-vars] False-positive on type guards that do not explicitly reference the variable

### ISSUE THREAD

**Issue by crunchyintheory** [typescript-eslint/typescript-eslint#2225]
**Repro**

Clone https://github.com/crunchyintheory/typescript-eslint-no-unused and run `npm i && npm test`

or

```JSON
{
    "rules": {
        "@typescript-eslint/no-unused-vars": ["error"]
    }
  }
```

```TS
const obj: Array<string> | boolean = new Array<string>();

const type: 'array' | 'boolean' = 'array';

function test(obj: Array<string> | boolean): obj is Array<string> { // 'obj' is defined but never used - eslint(@typescript-eslint/no-unused-vars)
    return type === 'array';
}

if(test(obj)) {
    // Typescript evaluates object as array
}
```

**Expected Result**

Type guards should count as a "use" of the variable

**Actual Result**

Type guards do not count the variable as used, and trigger an error unless the variable is explicitly referenced in the body of the function

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.3.0` |
| `@typescript-eslint/parser`        | `3.3.0` |
| `TypeScript`                       | `3.9.5` |
| `ESLint`                           | `7.2.0` |
| `node`                             | `12.18.1` |
| `npm`                              | `6.14.5` |

**Comment by bradzacher** [comment]
#1856
